test/model_server/model_server_test.cpp

```diff
@@ -0,0 +1,128 @@
+                       .SetUseCatalog(true)
+                       .SetWithModelServer(true)
+                       .SetUseNetwork(true)
+                       .SetUseGC(true)
+                       .SetUseExecution(true)
+                       .SetUseStatsStorage(true)
+                       .SetUseTrafficCop(true)
+                       .SetModelServerPath("../../script/model/model_server.py")
+                       .Build();
+
+    return db_main;
+  }
+
+  static bool GenerateMiniTrainerCSV() {
+    pid_t pid;
+    if ((pid = ::fork()) == 0) {
+      std::string mini_runner = MINI_RUNNER_PATH;
+      std::string bch_filter = "--benchmark_filter=SEQ0";
+      std::string row_limit = "--mini_runner_rows_limit=1000";
+      char *args[] = {mini_runner.data(), bch_filter.data(), row_limit.data(), nullptr};
+      if (::execvp(args[0], args) < 0) {
+        MODEL_SERVER_LOG_ERROR("failed to run {}: {}", MINI_RUNNER_PATH, strerror(errno));
+      }
+      return false;
+    } else if (pid > 0) {
+      ::waitpid(pid, NULL, 0);
+
+      /**
+       * TODO:
+       * Unfortunately, the current codebase relies on hardcoded name such as 'pipeline.csv' (mini_runners.cpp),
+       * and 'xxx_execution.csv' (mini_train.py) heavily.
+       * We would want to remove this hard-coded names in the future.
+       */
+      ::rename("pipeline.csv", TEST_FILE_NAME);
+      return true;
+    } else {
+      MODEL_SERVER_LOG_ERROR("failed to fork");
+    }
+    return false;
+  }
+
+  static void CleanUp(void) {
+    // Remove the file
+    ::remove(TEST_FILE_NAME);
+  }
+};
+
+// NOLINTNEXTLINE
+TEST_F(ModelServerTest, PipelineTest) {
+  messenger::messenger_logger->set_level(spdlog::level::info);
+  model_server_logger->set_level(spdlog::level::info);
+
+  // Generate the trace files
+  ASSERT_TRUE(GenerateMiniTrainerCSV());
+
+  auto primary = BuildDBMain();
+  primary->GetNetworkLayer()->GetServer()->RunServer();
+
+  auto ms_manager = primary->GetModelServerManager();
+
+  // Wait for the model server process to start
+  while (!ms_manager->ModelServerStarted()) {
+  }
+
+  std::vector<std::vector<double>> features{
+      {0, 10000, 4, 1, 10000, 1, 0, 0},
+      {0, 10000, 4, 1, 10000, 1, 0, 0},
+      {0, 10000, 4, 1, 10000, 1, 0, 0},
+      {0, 10000, 4, 1, 10000, 1, 0, 0},
+  };
+
+  // Send a message
+  std::string msg = "ModelServerTest";
+  ms_manager->PrintMessage(msg);
+
+  // Perform a training
+  std::vector<std::string> models{"lr"};
+  auto seq_files_dir = std::filesystem::current_path();
+  MODEL_SERVER_LOG_INFO("Training with {}, and input dir: {}", models[0], seq_files_dir);
+  ModelServerFuture<std::string> future;
+  ms_manager->TrainWith(models, seq_files_dir, &future);
+  auto res = future.Wait();
+  ASSERT_EQ(res.second, true);  // Training succeeds
+
+  // Perform inference
+  auto result = ms_manager->DoInference("OP_INTEGER_PLUS_OR_MINUS", features);
+  ASSERT_GT(result.size(), 0);
```

Comment:
Should't we check that the results have the same size as the features?

Reply:
Good idea!